File: backend/libbackend/password.ml
type t = string

let to_bytes pw = pw

let from_plaintext password =
  password
  |> Bytes.of_string
  |> Sodium.Password_hash.Bytes.wipe_to_password
  |> Sodium.Password_hash.Bytes.hash_password Sodium.Password_hash.interactive
  |> Bytes.to_string
  |> B64.encode


let invalid = ""

let from_hash pw = pw

File: backend/src/LibCloud/Password.fs
module LibBackend.Password

open Prelude

type T =
  private
  | Pw of string

  override this.ToString() : string =
    let (Pw pw) = this
    pw

let fromPlaintext (password : string) : T =
  password
  |> Sodium.PasswordHash.ArgonHashString
  |> UTF8.toBytes
  |> Base64.defaultEncodeToString
  |> Pw

let invalid : T = Pw ""

let fromHash (hash : string) : T = Pw hash
